runners/core-java/src/test/java/org/apache/beam/runners/core/StatefulDoFnRunnerTest.java

```diff
@@ -0,0 +1,188 @@
+import org.apache.beam.sdk.transforms.Sum;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.transforms.windowing.GlobalWindows;
+import org.apache.beam.sdk.transforms.windowing.IntervalWindow;
+import org.apache.beam.sdk.transforms.windowing.PaneInfo;
+import org.apache.beam.sdk.util.NullSideInputReader;
+import org.apache.beam.sdk.util.TimeDomain;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.util.WindowingStrategy;
+import org.apache.beam.sdk.util.state.StateSpec;
+import org.apache.beam.sdk.util.state.StateSpecs;
+import org.apache.beam.sdk.util.state.ValueState;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.TupleTag;
+import org.joda.time.Instant;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+/** Tests for {@link StatefulDoFnRunnerTest}. */
+@RunWith(JUnit4.class)
+public class StatefulDoFnRunnerTest {
+
+  @Mock StepContext mockStepContext;
+
+  @Mock TimerInternals mockTimerInternals;
+
+  private InMemoryLongSumAggregator droppedDueToLateness;
+  private AggregatorFactory aggregatorFactory;
+  private WindowingStrategy<?, ?> windowingStrategy;
+  private InMemoryStateInternals<String> stateInternals;
+
+  @Before
+  public void setup() {
+    MockitoAnnotations.initMocks(this);
+    when(mockStepContext.timerInternals()).thenReturn(mockTimerInternals);
+    droppedDueToLateness = new InMemoryLongSumAggregator("droppedDueToLateness");
+
+    aggregatorFactory = new AggregatorFactory() {
+      @Override
+      public <InputT, AccumT, OutputT> Aggregator<InputT, OutputT> createAggregatorForDoFn(
+          Class<?> fnClass, ExecutionContext.StepContext stepContext, String aggregatorName,
+          Combine.CombineFn<InputT, AccumT, OutputT> combine) {
+        return (Aggregator<InputT, OutputT>) droppedDueToLateness;
+      }
+    };
+    windowingStrategy = WindowingStrategy.of(new GlobalWindows());
+    stateInternals = new InMemoryStateInternals<>("hello");
+    when(mockStepContext.stateInternals()).thenReturn((StateInternals) stateInternals);
+  }
+
+  @Test
+  public void testLateDataDroping() {
+    when(mockTimerInternals.currentInputWatermarkTime())
+        .thenReturn(BoundedWindow.TIMESTAMP_MAX_VALUE);
+
+    DoFn<KV<String, Integer>, Integer> fn = new MyDoFn();
+
+    DoFnRunner<KV<String, Integer>, Integer> runner = DoFnRunners.statefulDoFnRunner(
+        fn, getSimpleDoFnRunner(fn), mockStepContext, aggregatorFactory, windowingStrategy);
+
+    runner.startBundle();
+
+    runner.processElement(
+        WindowedValue.of(KV.of("hello", 1), Instant.now(),
+            new IntervalWindow(new Instant(0), new Instant(10)), PaneInfo.NO_FIRING));
+
+    assertEquals(1L, droppedDueToLateness.sum);
+
+    runner.finishBundle();
+  }
+
+  @Test
+  public void testGc() {
```

Comment:
I think we should also have a test that adds state for several different windows/keys and checks whether GC works correctly.